### DISCUSSION TITLE
Schema doesn't allow for custom versioning regular expressions

### DISCUSSION THREAD

**Discussion by spilliams** [renovatebot/renovate#22247] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

When I use the schema file found [here](https://docs.renovatebot.com/renovate-schema.json) to check the syntax of my renovate.json file, my syntax checker gives me an error like the following:

```
Schema validation errors were encountered.
  renovate.json::$.terragrunt.versioning: 'regex:^(?<compatibility>.*)[/-](?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)$' is not one of ['aws-machine-image', 'azure-rest-api', 'bazel-module', 'cargo', 'composer', 'conan', 'deb', 'debian', 'docker', 'git', 'go-mod-directive', 'gradle', 'hashicorp', 'helm', 'hermit', 'hex', 'ivy', 'kubernetes-api', 'loose', 'maven', 'nixpkgs', 'node', 'npm', 'nuget', 'pep440', 'perl', 'poetry', 'python', 'redhat', 'regex', 'rez', 'ruby', 'semver', 'semver-coerced', 'swift', 'ubuntu']
```

I would like the schema file to reflect the fact that versioning strings like `regex:^((?<compatibility>.*)-v|v*)(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)$` are valid (this string is from [your own docs](https://docs.renovatebot.com/modules/manager/terragrunt/#additional-information))

### What you tried so far.

I have tried using the tool [check-jsonschema](https://check-jsonschema.readthedocs.io/en/latest/) v0.23.0 to check my renovate.json schema. This tool pulls your schema file from https://docs.renovatebot.com/renovate-schema.json per [this code](https://github.com/python-jsonschema/check-jsonschema/blob/23d783726457ff72efc9674e6eefc983147f434f/src/check_jsonschema/catalog.py#LL109C17-L109C66). When I tried this I received the above error.

**Comment by rarkins** [comment]
Here's the `versioning` definition in the JSON schema:

```
    "versioning": {
      "description": "Versioning to use for filtering and comparisons.",
      "type": "string",
      "enum": [
        "aws-machine-image",
        "azure-rest-api",
        "bazel-module",
        "cargo",
        "composer",
        "conan",
        "deb",
        "debian",
        "docker",
        "git",
        "go-mod-directive",
        "gradle",
        "hashicorp",
        "helm",
        "hermit",
        "hex",
        "ivy",
        "kubernetes-api",
        "loose",
        "maven",
        "nixpkgs",
        "node",
        "npm",
        "nuget",
        "pep440",
        "perl",
        "poetry",
        "python",
        "redhat",
        "regex",
        "rez",
        "ruby",
        "semver",
        "semver-coerced",
        "swift",
        "ubuntu"
      ]
    },
```

The enum is useful for most people, but it disallowed arbitrary regex:* strings. Is there any trick we can do with unions to allow it?

**Reply by viceice** [reply]
yes, we need a type with a string and pattern to be added with `oneOf` the enum

- https://json-schema.org/understanding-json-schema/reference/string.html#pattern
- https://json-schema.org/understanding-json-schema/reference/combining.html#oneof

**Comment by rarkins** [comment, marked as answer]
Created #22249
